### PR TITLE
[EmitC] Avoid emitting unused variables and empty struct definitions

### PR DESCRIPTION
**Pull request by simon-camp**

### CODE REVIEW

**Comment by simon-camp** [comment]
This should fix #8157, but I wasn't able to reproduce the error with my version of clang.

@powderluv Could you please try this out to see if this is all that's needed.

File: iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp [hunk 1]
```diff
@@ -1049,34 +1049,36 @@ LogicalResult createAPIFunctions(IREE::VM::ModuleOp moduleOp,
...
-        /*callee=*/StringAttr::get(ctx, "EMITC_STRUCT_PTR_MEMBER"),
-        /*args=*/
-        ArrayAttr::get(ctx, {builder.getIndexAttr(0),
-                             emitc::OpaqueAttr::get(ctx, "refs")}),
-        /*templateArgs=*/ArrayAttr{},
-        /*operands=*/ArrayRef<Value>{stateOp.getResult(0)});
-
-    for (int i = 0; i < numGlobalRefs; i++) {
-      auto refPtrOp = builder.create<emitc::CallOp>(
+    if (numGlobalRefs > 0) {
```
Comment (simon-camp) [inline iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp:1052]: The diff looks messy, but that guard is the only change.

**Comment by powderluv** [comment]
that fixed most of the errors but we still have 
```
/home/foo/github/iree/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp:1063:24: error: expression result unused [-Werror,-Wunused-value]
          /*location=*/loc,
                       ^~~
/home/foo/github/iree/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp:1068:73: error: use of undeclared identifier 'i'
              ctx, {builder.getIndexAttr(0), builder.getUI32IntegerAttr(i)}),
                                                                        ^
/home/foo/github/iree/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp:1078:40: error: use of undeclared identifier 'refPtrOp'
          /*operands=*/ArrayRef<Value>{refPtrOp.getResult(0)});
                                       ^
3 errors generated.
ninja: build stopped: subcommand failed.
```

**Comment by powderluv** [comment]
fyi clang version is latest from apt.llvm.org:

3995wx:~/github/iree$ clang --version
Ubuntu clang version 14.0.0-++20220112083514+453a056df616-1~exp1~20220112203528.227
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin

**Comment by simon-camp** [comment]
Okay sorry, it looks like you changed the code by hand and the comment above was misleading. The PR wraps these lines in the mentioned if statement:

https://github.com/google/iree/blob/11056ea22621d3aa666879ddef1c64fd19c59715/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp#L1052-L1080

**Comment by powderluv** [comment]
I am on the simon-camp:emitc-unused-variable branch now:

```
[0/2] Re-checking globbed directories...
[4165/4974] Building C object iree/vm/test/emitc/CMakeFiles/iree_vm_test_emitc_shift_ops.dir/__/__/module_impl_emitc.c.o
FAILED: iree/vm/test/emitc/CMakeFiles/iree_vm_test_emitc_shift_ops.dir/__/__/module_impl_emitc.c.o 

O3 -DNDEBUG -fPIC -fvisibility=hidden -Werror -Wall -Wno-ambiguous-member-template -Wno-char-subscripts -Wno-deprecated-declarations -Wno-extern-c-compat -Wno-gnu-alignof-expression -Wno-gnu-variable-sized-type-not-at-end -Wno-ignored-optimization-argument -Wno-invalid-offsetof -Wno-invalid-source-encoding -Wno-mismatched-tags -Wno-pointer-sign -Wno-reserved-user-defined-literal -Wno-return-type-c-linkage -Wno-self-assign-overloaded -Wno-sign-compare -Wno-signed-unsigned-wchar -Wno-strict-overflow -Wno-trigraphs -Wno-unknown-pragmas -Wno-unknown-warning-option -Wno-unused-command-line-argument -Wno-unused-const-variable -Wno-unused-function -Wno-unused-local-typedef -Wno-unused-private-field -Wno-user-defined-warnings -Wctad-maybe-unsupported -Wfloat-overflow-conversion -Wfloat-zero-conversion -Wfor-loop-analysis -Wformat-security -Wgnu-redeclared-enum -Wimplicit-fallthrough -Winfinite-recursion -Wliteral-conversion -Wnon-virtual-dtor -Woverloaded-virtual -Wself-assign -Wstring-conversion -Wtautological-overlap-compare -Wthread-safety -Wthread-safety-beta -Wunused-comparison -Wvla -DEMITC_IMPLEMENTATION=\"shift_ops.h\" -std=gnu11 -MD -MT iree/vm/test/emitc/CMakeFiles/iree_vm_test_emitc_shift_ops.dir/__/__/module_impl_emitc.c.o -MF iree/vm/test/emitc/CMakeFiles/iree_vm_test_emitc_shift_ops.dir/__/__/module_impl_emitc.c.o.d -o iree/vm/test/emitc/CMakeFiles/iree_vm_test_emitc_shift_ops.dir/__/__/module_impl_emitc.c.o -c /home/anush/github/iree/iree/vm/module_impl_emitc.c
In file included from /home/anush/github/iree/iree/vm/module_impl_emitc.c:7:
/home/anush/github/iree/build/iree/vm/test/emitc/shift_ops.h:64:40: error: variable 'v10' set but not used [-Werror,-Wunused-but-set-variable]
  const shift_ops_test_shl_i32_args_t* v10;
                                       ^
/home/anush/github/iree/build/iree/vm/test/emitc/shift_ops.h:67:36: error: variable 'v13' set but not used [-Werror,-Wunused-but-set-variable]
  shift_ops_test_shl_i32_result_t* v13;
                                   ^
/home/anush/github/iree/build/iree/vm/test/emitc/shift_ops.h:160:41: error: variable 'v10' set but not used [-Werror,-Wunused-but-set-variable]
  const shift_ops_test_shr_i32s_args_t* v10;
                                        ^
/home/anush/github/iree/build/iree/vm/test/emitc/shift_ops.h:163:37: error: variable 'v13' set but not used [-Werror,-Wunused-but-set-variable]
  shift_ops_test_shr_i32s_result_t* v13;
                                    ^
/home/anush/github/iree/build/iree/vm/test/emitc/shift_ops.h:254:41: error: variable 'v10' set but not used [-Werror,-Wunused-but-set-variable]
  const shift_ops_test_shr_i32u_args_t* v10;
                                        ^
/home/anush/github/iree/build/iree/vm/test/emitc/shift_ops.h:257:37: error: variable 'v13' set but not used [-Werror,-Wunused-but-set-variable]
  shift_ops_test_shr_i32u_result_t* v13;
                                    ^
/home/anush/github/iree/build/iree/vm/test/emitc/shift_ops.h:361:18: error: variable 'v12' set but not used [-Werror,-Wunused-but-set-variable]
  iree_vm_ref_t* v12;
                 ^
/home/anush/github/iree/build/iree/vm/test/emitc/shift_ops.h:362:11: error: variable 'v13' set but not used [-Werror,-Wunused-but-set-variable]
  int32_t v13;
          ^
8 errors generated.
[4166/4974] Building C object iree/vm/test/emitc/CMakeFiles/iree_vm_test_emitc_ref_ops.dir/__/__/module_impl_emitc.c.o
/usr/bin/clang -DIREE_HAVE_EMITC_DIALECT -I/home/anush/github/iree/build/iree/vm/test/emitc -isystem /home/anush/github/iree -isystem /home/anush/github/iree/build -O3 -DNDEBUG -fPIC -fvisibility=hidden -Werror -Wall -Wno-ambiguous-member-template -Wno-char-subscripts -Wno-deprecated-declarations -Wno-extern-c-compat -Wno-gnu-alignof-expression -Wno-gnu-variable-sized-type-not-at-end -Wno-ignored-optimization-argument -Wno-invalid-offsetof -Wno-invalid-source-encoding -Wno-mismatched-tags -Wno-pointer-sign -Wno-reserved-user-defined-literal -Wno-return-type-c-linkage -Wno-self-assign-overloaded -Wno-sign-compare -Wno-signed-unsigned-wchar -Wno-strict-overflow -Wno-trigraphs -Wno-unknown-pragmas -Wno-unknown-warning-option -Wno-unused-command-line-argument -Wno-unused-const-variable -Wno-unused-function -Wno-unused-local-typedef -Wno-unused-private-field -Wno-user-defined-warnings -Wctad-maybe-unsupported -Wfloat-overflow-conversion -Wfloat-zero-conversion -Wfor-loop-analysis -Wformat-security -Wgnu-redeclared-enum -Wimplicit-fallthrough -Winfinite-recursion -Wliteral-conversion -Wnon-virtual-dtor -Woverloaded-virtual -Wself-assign -Wstring-conversion -Wtautological-overlap-compare -Wthread-safety -Wthread-safety-beta -Wunused-comparison -Wvla -DEMITC_IMPLEMENTATION=\"ref_ops.h\" -std=gnu11 -MD -MT iree/vm/test/emitc/CMakeFiles/iree_vm_test_emitc_ref_ops.dir/__/__/module_impl_emitc.c.o -MF iree/vm/test/emitc/CMakeFiles/iree_vm_test_emitc_ref_ops.dir/__/__/module_impl_emitc.c.o.d -o iree/vm/test/emitc/CMakeFiles/iree_vm_test_emitc_ref_ops.dir/__/__/module_impl_emitc.c.o -c /home/anush/github/iree/iree/vm/module_impl_emitc.c
In file included from /home/anush/github/iree/iree/vm/module_impl_emitc.c:7:
/home/anush/github/iree/build/iree/vm/test/emitc/ref_ops.h:67:42: error: variable 'v10' set but not used [-Werror,-Wunused-but-set-variable]
  const ref_ops_test_zero_ref_eq_args_t* v10;
                                         ^
/home/anush/github/iree/build/iree/vm/test/emitc/ref_ops.h:70:38: error: variable 'v13' set but not used [-Werror,-Wunused-but-set-variable]
  ref_ops_test_zero_ref_eq_result_t* v13;
                                     ^
/home/anush/github/iree/build/iree/vm/test/emitc/ref_ops.h:218:37: error: variable 'v10' set but not used [-Werror,-Wunused-but-set-variable]
  const ref_ops_test_ref_ne_args_t* v10;
                                    ^
/home/anush/github/iree/build/iree/vm/test/emitc/ref_ops.h:221:33: error: variable 'v13' set but not used [-Werror,-Wunused-but-set-variable]
  ref_ops_test_ref_ne_result_t* v13;
                                ^
/home/anush/github/iree/build/iree/vm/test/emitc/ref_ops.h:378:37: error: variable 'v10' set but not used [-Werror,-Wunused-but-set-variable]
  const ref_ops_test_ref_nz_args_t* v10;
                                    ^
/home/anush/github/iree/build/iree/vm/test/emitc/ref_ops.h:381:33: error: variable 'v13' set but not used [-Werror,-Wunused-but-set-variable]
  ref_ops_test_ref_nz_result_t* v13;
                                ^
/home/anush/github/iree/build/iree/vm/test/emitc/ref_ops.h:527:11: error: variable 'v25' set but not used [-Werror,-Wunused-but-set-variable]
  int32_t v25;
          ^
/home/anush/github/iree/build/iree/vm/test/emitc/ref_ops.h:526:18: error: variable 'v24' set but not used [-Werror,-Wunused-but-set-variable]
  iree_vm_ref_t* v24;
                 ^
8 errors generated.
[4187/4974] Building C object iree/vm/test/emitc/CMakeFiles/iree_vm_test_emitc_global_ops_f32.dir/__/__/module_impl_emitc.c.o
FAILED: iree/vm/test/emitc/CMakeFiles/iree_vm_test_emitc_global_ops_f32.dir/__/__/module_impl_emitc.c.o 
/usr/bin/clang -DIREE_HAVE_EMITC_DIALECT -I/home/anush/github/iree/build/iree/vm/test/emitc -isystem /home/anush/github/iree -isystem /home/anush/github/iree/build -O3 -DNDEBUG -fPIC -fvisibility=hidden -Werror -Wall -Wno-ambiguous-member-template -Wno-char-subscripts -Wno-deprecated-declarations -Wno-extern-c-compat -Wno-gnu-alignof-expression -Wno-gnu-variable-sized-type-not-at-end -Wno-ignored-optimization-argument -Wno-invalid-offsetof -Wno-invalid-source-encoding -Wno-mismatched-tags -Wno-pointer-sign -Wno-reserved-user-defined-literal -Wno-return-type-c-linkage -Wno-self-assign-overloaded -Wno-sign-compare -Wno-signed-unsigned-wchar -Wno-strict-overflow -Wno-trigraphs -Wno-unknown-pragmas -Wno-unknown-warning-option -Wno-unused-command-line-argument -Wno-unused-const-variable -Wno-unused-function -Wno-unused-local-typedef -Wno-unused-private-field -Wno-user-defined-warnings -Wctad-maybe-unsupported -Wfloat-overflow-conversion -Wfloat-zero-conversion -Wfor-loop-analysis -Wformat-security -Wgnu-redeclared-enum -Wimplicit-fallthrough -Winfinite-recursion -Wliteral-conversion -Wnon-virtual-dtor -Woverloaded-virtual -Wself-assign -Wstring-conversion -Wtautological-overlap-compare -Wthread-safety -Wthread-safety-beta -Wunused-comparison -Wvla -DEMITC_IMPLEMENTATION=\"global_ops_f32.h\" -std=gnu11 -MD -MT iree/vm/test/emitc/CMakeFiles/iree_vm_test_emitc_global_ops_f32.dir/__/__/module_impl_emitc.c.o -MF iree/vm/test/emitc/CMakeFiles/iree_vm_test_emitc_global_ops_f32.dir/__/__/module_impl_emitc.c.o.d -o iree/vm/test/emitc/CMakeFiles/iree_vm_test_emitc_global_ops_f32.dir/__/__/module_impl_emitc.c.o -c /home/anush/github/iree/iree/vm/module_impl_emitc.c
In file included from /home/anush/github/iree/iree/vm/module_impl_emitc.c:7:
/home/anush/github/iree/build/iree/vm/test/emitc/global_ops_f32.h:62:53: error: variable 'v10' set but not used [-Werror,-Wunused-but-set-variable]
  const global_ops_f32_test_global_load_f32_args_t* v10;
                                                    ^
/home/anush/github/iree/build/iree/vm/test/emitc/global_ops_f32.h:65:49: error: variable 'v13' set but not used [-Werror,-Wunused-but-set-variable]
  global_ops_f32_test_global_load_f32_result_t* v13;
                                                ^
/home/anush/github/iree/build/iree/vm/test/emitc/global_ops_f32.h:108:54: error: variable 'v10' set but not used [-Werror,-Wunused-but-set-variable]
  const global_ops_f32_test_global_store_f32_args_t* v10;
                                                     ^
/home/anush/github/iree/build/iree/vm/test/emitc/global_ops_f32.h:111:50: error: variable 'v13' set but not used [-Werror,-Wunused-but-set-variable]
  global_ops_f32_test_global_store_f32_result_t* v13;
                                                 ^
/home/anush/github/iree/build/iree/vm/test/emitc/global_ops_f32.h:216:18: error: variable 'v12' set but not used [-Werror,-Wunused-but-set-variable]
  iree_vm_ref_t* v12;
                 ^
/home/anush/github/iree/build/iree/vm/test/emitc/global_ops_f32.h:217:11: error: variable 'v13' set but not used [-Werror,-Wunused-but-set-variable]
  int32_t v13;
          ^
6 errors generated.
[4294/4974] Building CXX object iree/hal/vmvx/cts/CMakeFiles/iree_hal_vmvx_cts_vmvx_command_buffer_test.dir/command_buffer_vmvx_test.cc.o
ninja: build stopped: subcommand failed.

```

**Review by powderluv — APPROVED** [review]
verified it fixes the issues. thx

**Review by benvanik — APPROVED** [review]

**Pull request merged**